std/experimental/color.d

```diff
@@ -0,0 +1,704 @@
+module std.experimental.color;
+
+import std.traits: isFloatingPoint, isIntegral, isSigned, isSomeChar;
+
+
+// *** std.algorithm is not @nogc or nothrow >_<
+private
+{
+    bool allIn(Range)(Range things, Range range) if (isInputRange!Range)
+    {
+        foreach(thing; things)
+            if(!range.canFind(thing))
+                return false;
+        return true;
+    }
+    bool anyIn(Range)(Range things, Range range) if (isInputRange!Range)
+    {
+        foreach(thing; things)
+            if(range.canFind(thing))
+                return true;
+        return false;
+    }
+
+    ElementType!Range notIn(Range)(Range things, Range range) if (isInputRange!Range)
+    {
+        foreach(thing; things)
+            if(!range.canFind(thing))
+                return thing;
+        return true;
+    }
+}
+
+@safe: pure: nothrow: @nogc:
+
+// defined color spaces, more to come!
+enum ColorSpace
+{
+    XYZ,        // CIE 1931
+//    UVW,        // CIE 1964
+//    Lab,        // CIA Lab
+
+    sRGB,       // sRGB
+    Gamma,      // Gamma 2.2 sRGB
+    Linear,     // Linear sRGB
+
+//    AdobeRGB,   // Adobe RGB
+//    AdobeLinear,// Linear Adobe RGB
+
+//    YUV,        // There are multiple YUV spaces...
+
+//    HSV,        // HSV and HSL?
+
+//    CMYK        // Maybe not; CMYK depends on OS color profiles from printer drivers and stuff...
+}
+
+enum isColor(T) = is(T == Color!(cmp, CT, cs), string cmp, CT, ColorSpace cs);
+enum isRGB(ColorSpace c) = c == ColorSpace.sRGB || c == ColorSpace.Gamma || c == ColorSpace.Linear;
+
+struct Color(string components_, ComponentType_, ColorSpace colorSpace_ = ColorSpace.sRGB) if(isValidComponentType!ComponentType_)
+{
+    // mixin to fabricate component members
+    private static string declComponents()
+    {
+        string s;
+        foreach(i, c; components)
+            s ~= (c == 'x' ? "private " : "") ~ ComponentType.stringof ~ ' ' ~ c ~ ";\n";
+        return s;
+    }
+    mixin(declComponents());
+
+@safe: pure: nothrow: @nogc:
+
+    alias ComponentType = ComponentType_;
+    enum string components = components_;
+    enum ColorSpace colorSpace = colorSpace_;
```

Comment:
I don't think it can hurt for them to be public. People might want to know the colour space, and the component type is definitely interesting. I could probably make components private.
Is there an idiomatic solution to this problem?